File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys

input = sys.stdin.readline

N, C = map(int, input().split())

home = [int(input()) for _ in range(N)]
home.sort()

start = 1
end = home[-1] - home[0]
answer = 0

while start <= end:
    mid = (start + end) // 2
    current_home = home[0]
    cnt = 1

    for i in range(1, len(home)):
        if home[i] >= current_home + mid:
            cnt += 1
            current_home = home[i]

    if cnt >= C:
        start = mid + 1
        answer = mid
    elif cnt < C:
        end = mid - 1

print(answer)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());

        ArrayList<Integer> homes = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            homes.add(Integer.parseInt(st.nextToken()));
        }
        Collections.sort(homes);
        int start = 1;
        int end = homes.get(N - 1) - homes.get(0);
        int answer = 0;

        while (start <= end) {
            int mid = (start + end) / 2;
            int current_home = homes.get(0);
            int cnt = 1;
            for (int i = 1; i < N; i++) {
                if (homes.get(i) >= current_home + mid) {

                    cnt += 1;
                    current_home = homes.get(i);
                }
            }

            if (cnt >= C) {
                start = mid + 1;
                answer = mid;
            } else {
                end = mid - 1;
            }

        }
        System.out.println(answer);
    }
}